---
title: "Chimclustr report"
output: html_document
---

```{r setup, echo=FALSE, message = FALSE, warning = FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
stopifnot(require(tidyverse),
          require(pheatmap),
          require(cowplot))
stopifnot(exists("plot_1_data"),
          exists("plot_2_data"),
          exists("sil_data"),
          exists("read_allele_flt"),
          exists("hap_em_search"))
```

### Pre-filter Read-Allele Heatmap
```{r plot_1, fig.width=8, fig.height=6}
pheatmap(mat = plot_1_data$mat,
         cluster_rows = plot_1_data$row_hclust,
         cluster_cols = FALSE,
         show_colnames = FALSE,
         show_rownames = FALSE,
         annotation_row = plot_1_data$row_annot,
         annotation_col = plot_1_data$col_annot,
         annotation_colors = list(status = c(pass = 'chartreuse4', fail = 'darkorange4')),
         color = viridisLite::cividis(2),
         na_col = 'gray50',
         legend = FALSE)
```

### PAM-k clustering Read-Allele heatmap
```{r plot_2, fig.width=8, fig.height=6}
pheatmap(mat = plot_2_data$mat,
         cluster_rows = plot_2_data$row_hclust,
         cluster_cols = FALSE,
         show_colnames = FALSE,
         show_rownames = FALSE,
         color = viridisLite::cividis(2),
         annotation_row = plot_2_data$row_annot,
         annotation_colors = plot_2_data$annot_color,
         annotation_legend = FALSE,
         na_col = 'gray50',
         legend = FALSE)
```

```{r sil_plots, fig.width=8, fig.height=6}
sil_data %>%
  filter(k > 1) %>%
  select(k, cluster, rid, sil_width, is_noise, k_asw) %>%
  mutate(lab = str_c('k = ', k,', asw = ', format(k_asw, digits = 2))) %>%
  unnest(c(rid, sil_width)) %>%
  arrange(k, desc(cluster), sil_width) %>%
  mutate(kcr = str_c(k, cluster, rid, sep = '_') %>% as_factor()) %>%
  ggplot(aes(kcr, sil_width, fill = cluster, col = cluster)) +
  coord_flip() +
  geom_col() +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(),
        axis.title.y = element_blank()) +
  facet_wrap(~lab, scales = 'free_y') +
  ggtitle('PAM-k clustering silhouette plots')
```

```{r sil_plots_2, fig.width=8, fig.height=6}
pam_k_asw <-
  sil_data %>%
  filter(k > 1) %>%
  select(k, k_asw, k_asw_filt) %>%
  distinct()

sil_data %>%
  filter(k > 1) %>%
  select(k, cluster, sil_width, is_noise, k_asw, k_asw_filt) %>%
  unnest(sil_width) %>%
  ggplot(aes(cluster, sil_width)) +
  geom_boxplot(aes(fill = is_noise), col = 'gray25') +
  geom_hline(aes(linetype = 'max asw w/o noise', yintercept = max(pam_k_asw$k_asw_filt)),
             col = 'red') +
  geom_hline(data = pam_k_asw, aes(yintercept = k_asw, linetype = 'asw')) +
  geom_hline(data = pam_k_asw, aes(yintercept = k_asw/2, linetype = 'noise threshold'),
             col='darkblue') +
  geom_hline(data = pam_k_asw, aes(yintercept = k_asw_filt, linetype = 'asw w/o noise')) +
  scale_fill_manual(name = 'noise', values = c(`TRUE` = 'gray50', `FALSE` = 'gray80')) +
  scale_linetype_manual(name = NULL, values = c(1,2,1,1),
                        guide = guide_legend(override.aes = list(color = c("black", "black", "red", 'darkblue')))) +
  facet_grid(~k, scales = 'free_x', space = 'free_x', labeller = 'label_both') +
  ggtitle('PAM-k clustering silhouette width boxplot')
```

```{r make_plot_lists}
em_plots <-
  hap_em_search %>%
  pmap(function(k, label, em_res, LH, ...) {
    p_lh <-
      em_res %>%
      select(iter, LH) %>%
      ggplot(aes(iter, LH)) +
      geom_line() +
      geom_point(size = 0.5) +
      ylab('log likelihood') +
      xlab('iteration')
    
    p_ve <-
      em_res %>%
      select(iter, rate = var_error_rate) %>%
      mutate(rate = map(rate, ~ tibble(rate = ., id = seq_along(.)))) %>%
      unnest(rate) %>%
      mutate(id = as_factor(id)) %>%
      ggplot(aes(iter, rate, group = id)) +
      geom_line(alpha = 0.25) +
      ylab('variant error rate') +
      xlab('iteration')
    
    p_re <-
      em_res %>%
      select(iter, rate = read_error_rate) %>%
      mutate(rate = map(rate, ~ tibble(rate = ., id = seq_along(.)))) %>%
      unnest(rate) %>%
      mutate(id = as_factor(id)) %>%
      ggplot(aes(iter, rate, group = id)) +
      geom_line(alpha = 0.10) +
      ylab('read error rate') +
      xlab('iteration')
    
    p_post <-
      last(em_res$read_hap_post) %>%
      group_by(read_id) %>%
      arrange(desc(post_lh)) %>%
      slice(1) %>%
      ungroup() %>%
      select(read_id, haplotype) %>%
      left_join(select(em_res, iter, read_hap_post) %>% unnest(read_hap_post),
                by = c('read_id', 'haplotype')) %>%
      ggplot(aes(iter, post_lh, group = read_id)) +
      geom_line(alpha = 0.10) +
      ylab('posterior probability') +
      xlab('iteration') +
      facet_wrap(~haplotype, ncol = 1)
    
    p1 <- plot_grid(p_lh, p_ve, p_re, align = 'v', ncol = 1)
    p2 <- plot_grid(p1, p_post, ncol = 2)
    plot_grid(ggdraw() + draw_text(str_c('Haplotype Mixture EM Inspection (', label, ')')), p2,
              ncol = 1, rel_heights = c(1,15))
  })

phase_plots <-
  hap_em_search %>%
  pmap(function(k, label, em_res, LH, ...) {
    read_data <-
      last(em_res$read_hap_post) %>%
      group_by(read_id) %>%
      arrange(desc(post_lh)) %>%
      slice(1) %>%
      ungroup() %>%
      arrange(read_id) %>%
      mutate(error_rate = last(em_res$read_error_rate)) %>%
      select(read_id, haplotype, post_lh, error_rate) %>%
      mutate(status = if_else(error_rate < 0.05 & post_lh > 0.99 & haplotype != 'chimera', 'pass', 'fail')) %>%
      arrange(desc(haplotype), status, desc(error_rate), post_lh) %>%
      mutate(read_id = as_factor(as.character(read_id)))
    
    p1 <-
      unname(t(read_allele_flt)) %>%
      as_tibble() %>%
      mutate(read_id = seq_len(n())) %>%
      pivot_longer(starts_with('V'),
                   names_to = 'variant',
                   names_prefix = 'V',
                   names_ptypes = list(variant = integer()),
                   values_to = 'allele') %>%
      mutate(read_id = factor(read_id, levels(read_data$read_id)),
             allele = as.factor(allele),
             variant = as_factor(variant)) %>%
      left_join(select(read_data, haplotype, status, read_id), 'read_id') %>%
      ggplot(aes(variant, read_id, fill = allele)) +
      geom_tile() +
      scale_fill_manual(values = setNames(viridisLite::cividis(2), 1:2),
                        na.value = 'gray50') +
      facet_grid(haplotype ~ ., space='free_y', scales = 'free_y', switch = 'both') +
      theme(axis.ticks = element_blank(), axis.text = element_blank(),
            axis.title.y = element_blank(), legend.position = 'top',
            plot.margin = margin(t = 0, r = 0, b = 0, l = 0, unit = "pt"))
    
    p2 <-
      read_data %>%
      ggplot(aes(x='x', read_id, fill = status)) +
      geom_tile() +
      theme() +
      facet_grid(haplotype ~ ., space='free_y', scales = 'free_y', switch = 'both') +
      theme(legend.position = 'top', strip.background = element_blank(), strip.text = element_blank(),
            axis.ticks = element_blank(), axis.text = element_blank(), axis.title.y = element_blank(),
            plot.margin = margin(t = 0, r = 0, b = 0, l = 0, unit = "pt")) +
      xlab('status')
    
    p3 <-
      read_data %>%
      ggplot(aes(x='x', read_id, fill = error_rate)) +
      geom_tile() +
      theme() +
      scale_fill_viridis_c(limits = c(0,1)) +
      facet_grid(haplotype ~ ., space='free_y', scales = 'free_y', switch = 'both') +
      theme(legend.position = 'top', strip.background = element_blank(), strip.text = element_blank(),
            axis.ticks = element_blank(), axis.text = element_blank(), axis.title.y = element_blank(),
            plot.margin = margin(t = 0, r = 0, b = 0, l = 0, unit = "pt")) +
      xlab('error')
    
    p4 <-
      read_data %>%
      rename(posterior = post_lh) %>%
      ggplot(aes(x='x', read_id, fill = posterior)) +
      geom_tile() +
      theme() +
      scale_fill_viridis_c(option = 'plasma', limits = c(0, 1)) +
      facet_grid(haplotype ~ ., space='free_y', scales = 'free_y', switch = 'both') +
      theme(legend.position = 'top', strip.background = element_blank(), strip.text = element_blank(),
            axis.ticks = element_blank(), axis.text = element_blank(), axis.title.y = element_blank(),
            plot.margin = margin(t = 0, r = 0, b = 0, l = 0, unit = "pt")) +
      xlab('posterior')
    
    rm_legend <- theme(legend.position = 'none')
    
    title <- ggdraw() + draw_text(str_c('Haplotype Mixture EM Results (', label, ')'))
    pt <- plot_grid(get_legend(p1), get_legend(p2), get_legend(p3), get_legend(p4), nrow = 1)
    pb <- plot_grid(p1 + rm_legend, p2 + rm_legend, p3 + rm_legend, p4 + rm_legend, nrow = 1, rel_widths = c(10,1,1,1), align = 'h', axis = 'bt')
    plot_grid(title, pt, pb, ncol = 1, rel_heights = c(1, 1, 10))
  })
```

```{r em_plots, fig.width=8, fig.height=8}
seq_along(em_plots) %>% 
  walk(function(i) {
    print(em_plots[[i]])
    print(phase_plots[[i]])
  })
```

